csharp/src/Signer.cs

```diff
@@ -132,27 +153,79 @@ public Signer Body(byte[] body)
-            var jwsHeaders = new Dictionary<string, object>()
+        public TSigner Body(string body) => Body(body.ToUtf8());
+        
+        protected internal Dictionary<string, object> CreateJwsHeaders() =>
+            new()
             {
                 {"alg", "ES512"},
-                {"kid", kid},
+                {"kid", _kid},
                 {"tl_version", "2"},
-                {"tl_headers", string.Join(",", headerList.Select(h => h.Key))},
+                {"tl_headers", string.Join(",", _headers.Select(h => h.Key))},
             };
-            var signingPayload = Util.BuildV2SigningPayload(method, path, headerList, body);
+    }
+    
+    /// <summary>
+    /// Builder to generate a Tl-Signature header value.
+    /// </summary>
+    public abstract class AsyncSigner : Signer<AsyncSigner>
+    {
+        protected internal AsyncSigner(string kid) : base(kid)
+        {
+        }
+
+        /// <summary>Produce a JWS `Tl-Signature` v2 header value.</summary>
+        public abstract Task<string> SignAsync();
+    }
+
+    internal sealed class PrivateKeySigner : Signer
+    {
+        private readonly ECDsa _key;
+
+        internal PrivateKeySigner(string kid, ECDsa privateKey) : base(kid)
+        {
+            _key = privateKey;
+        }
+        
+        public override string Sign()
+        {
+            var jwsHeaders = CreateJwsHeaders();
+            var headerList = _headers.Select(e => (e.Key, e.Value)).ToList();
+            var signingPayload = Util.BuildV2SigningPayload(_method, _path, headerList, _body);
 
             return JWT.EncodeBytes(
                 signingPayload,
-                key,
+                _key,
                 JwsAlgorithm.ES512,
                 jwsHeaders,
-                options: new JwtOptions { DetachPayload = true });
+                options: new JwtOptions {DetachPayload = true});
+        }
+    }
+    
+    internal sealed class FunctionSigner : AsyncSigner
+    {
+        private readonly Func<string, Task<string>> _signAsync;
+        
+        internal FunctionSigner(string kid, Func<string, Task<string>> signAsync) : base(kid)
+        {
+            _signAsync = signAsync;
+        }
+        
+        public override async Task<string> SignAsync()
+        {
+            var jwsHeaders = CreateJwsHeaders();
+            var serializedJwsHeaders = JsonSerializer.SerializeToUtf8Bytes(jwsHeaders);
+            var serializedJwsHeadersB64 = Base64Url.Encode(serializedJwsHeaders);
+
+            var headerList = _headers.Select(e => (e.Key, e.Value)).ToList();
+            var signingPayload = Util.BuildV2SigningPayload(_method, _path, headerList, _body);
+            var signingPayloadB64 = Base64Url.Encode(signingPayload);
+
+            var signingMessage = $"{serializedJwsHeadersB64}.{signingPayloadB64}";
+
+            var signature = await _signAsync(signingMessage);
+
+            return $"{serializedJwsHeadersB64}..{signature}";
```

Comment:
Is this additional `dot` intentional? For clients to insert the _body in the middle?

Reply:
Yes here it's intentional. The standard JWS has the format `<headers>.<payload>.<signature>`, and the detached JWS which we use for our signatures have the payload removed, leaving the format `<headers>..<signature>`.